File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.PriorityQueue;
import java.util.Collections;


class Solution {
    public int[] solution(String[] operations) {
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        
        for(String operation : operations) {
            String[] op = operation.split(" ");

            if("I".equals(op[0])) {
                minHeap.add(Integer.valueOf(op[1]));
                maxHeap.add(Integer.valueOf(op[1]));
            } else if("D".equals(op[0])) {
                if("1".equals(op[1])) {
                    minHeap.remove(maxHeap.poll());
                } else if("-1".equals(op[1])) {
                    maxHeap.remove(minHeap.poll());
                }
            }
        }
                            
        Integer minValue = minHeap.poll();
        Integer maxValue = maxHeap.poll();

        return new int []{
            (maxValue == null) ? 0 : maxValue,
            (minValue == null) ? 0 : minValue
        };
    }
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.kt
import java.util.PriorityQueue
import java.util.Collections

class Solution {
    fun solution(operations: Array<String>): IntArray {
        val minHeap: PriorityQueue<Int> = PriorityQueue()
        val maxHeap: PriorityQueue<Int> = PriorityQueue(Collections.reverseOrder())
        
        operations
            .map { it.split(" ") }
            .forEach { op -> 
                when(op[0]) {
                    "I" -> {
                        minHeap.add(Integer.valueOf(op[1]))
                        maxHeap.add(Integer.valueOf(op[1]))
                    }

                    "D" -> {
                        when(op[1]) {
                            "1" -> minHeap.remove(maxHeap.poll())
                            "-1" -> maxHeap.remove(minHeap.poll())
                        }
                    }
                }
            }
        return intArrayOf(
            maxHeap.poll() ?: 0,
            minHeap.poll() ?: 0
        )
    }
}